File: Data Structures & Algorithms/same-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        # Base Case
        if not p and not q:
            return True
        if not p or not q:
            return False
        
        # Recursive Conditon
        if p.val != q.val:
            return False

        # Recursive Calls
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */



public class Solution {
    int kthSmallestValue = -1;
    int currentPosition = 0;

    public int KthSmallest(TreeNode root, int k) {
        dfsHelper(root, k);
        return kthSmallestValue;
    }

    // in order traversal 
    public void dfsHelper(TreeNode root, int k) {
        // base case 
        if (root == null || kthSmallestValue != -1) {
            return;
        }

        // process the left subtree
        dfsHelper(root.left, k);

        // process the current node 
        currentPosition += 1; 

        if (currentPosition == k) {
            kthSmallestValue = root.val;
            return;
        }

        // process the right subtrree
        dfsHelper(root.right, k);
    }
}
